crates/latch_core/src/ecs/storage/macros.rs

```diff
@@ -0,0 +1,141 @@
+// macros.rs - Macros for accessing component columns
+//
+// These macros provide a safe, ergonomic interface for accessing multiple
+// component slices simultaneously from archetype storage.
+//
+// # Safety
+//
+// The macros ensure safety by:
+// 1. Verifying all component IDs are unique at runtime
+// 2. Getting separate column pointers that don't alias
+// 3. Respecting the double-buffer read/write separation
+//
+// This allows safe parallel access to different component types while
+// preventing undefined behavior from aliasing mutable references.
+//
+// # Example
+//
+// ```ignore
+// use latch_core::{columns, columns_mut};
+//
+// // Read from current buffer
+// let positions = columns!(storage, Position);
+//
+// // Write to next buffer
+// let velocities = columns_mut!(storage, Velocity);
+//
+// // Read multiple (current buffer)
+// let (pos, vel) = columns!(storage, Position, Velocity);
+//
+// // Write multiple (next buffer)
+// let (pos_out, vel_out) = columns_mut!(storage, Position, Velocity);
+// ```
+
+/// Macro to get multiple immutable component slices from a storage.
+/// 
+/// Reads from the "current" buffer (stable state from last tick).
+/// Handles any number of components using compile-time validation.
+/// 
+/// # Example
+/// ```ignore
+/// let positions = columns!(storage, Position);
+/// let (positions, velocities) = columns!(storage, Position, Velocity);
+/// let (a, b, c) = columns!(storage, ComponentA, ComponentB, ComponentC);
+/// ```
+#[macro_export]
+macro_rules! columns {
+    // Single component - just call the method directly
+    ($storage:expr, $T:ty) => {
+        $storage.column_as_slice::<$T>().unwrap()
+    };
+    
+    // Multiple components - use the general implementation
+    ($storage:expr, $($T:ty),+ $(,)?) => {{
+        // Collect component IDs and verify uniqueness at compile time
+        let ids = [$(<$T as $crate::ecs::Component>::ID),+];
+        
+        // Get the current buffer index (what we read from)
+        let current_buffer = $storage.current_buffer_index();
+        
+        // Get raw pointers to each column
+        let ptrs = [$(unsafe {
+            $storage.get_column_ptr_const(<$T as $crate::ecs::Component>::ID)
+                .expect("Component not found in archetype")
+        }),+];
+        
+        // Runtime verification that all IDs are unique (not strictly necessary for immutable, but consistent)
+        for i in 0..ids.len() {
+            for j in (i+1)..ids.len() {
+                assert_ne!(ids[i], ids[j], "Duplicate component IDs in query");
+            }
+        }
```

Comment:
For immutable slices, is this runtime check necessary?

I get that it's low-cost, but you can have unlimited immutable borrows of a value. It's only mutable borrows that we have to ensure are unique.